stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -334,108 +365,514 @@ impl NakamotoBlockHeader {
+            burn_spent,
+            consensus_hash,
+            parent_block_id,
+            tx_merkle_root: Sha512Trunc256Sum([0u8; 32]),
+            state_index_root: TrieHash([0u8; 32]),
+            miner_signature: MessageSignature::empty(),
+            stacker_signature: MessageSignature::empty(),
+        }
+    }
+
+    /// Make a completely empty header
+    pub fn empty() -> NakamotoBlockHeader {
+        NakamotoBlockHeader {
+            version: 0,
+            chain_length: 0,
+            burn_spent: 0,
+            consensus_hash: ConsensusHash([0u8; 20]),
+            parent_block_id: StacksBlockId([0u8; 32]),
+            tx_merkle_root: Sha512Trunc256Sum([0u8; 32]),
+            state_index_root: TrieHash([0u8; 32]),
+            miner_signature: MessageSignature::empty(),
+            stacker_signature: MessageSignature::empty(),
+        }
+    }
+
+    /// Make a genesis header (testing only)
+    pub fn genesis() -> NakamotoBlockHeader {
+        NakamotoBlockHeader {
+            version: 0,
+            chain_length: 0,
+            burn_spent: 0,
+            consensus_hash: FIRST_BURNCHAIN_CONSENSUS_HASH.clone(),
+            parent_block_id: StacksBlockId(BOOT_BLOCK_HASH.0.clone()),
+            tx_merkle_root: Sha512Trunc256Sum([0u8; 32]),
+            state_index_root: TrieHash([0u8; 32]),
+            miner_signature: MessageSignature::empty(),
+            stacker_signature: MessageSignature::empty(),
+        }
+    }
 }
 
 impl NakamotoBlock {
-    /// Did the stacks tenure change on this nakamoto block? i.e., does this block
-    ///  include a TenureChange transaction?
-    pub fn tenure_changed(&self, parent: &StacksBlockId) -> bool {
+    /// Find all positionally-valid tenure changes in this block.
+    /// They must be the first transactions.
+    /// Return their indexes into self.txs
+    fn find_tenure_changes(&self) -> Vec<usize> {
+        let mut ret = vec![];
+        for (i, tx) in self.txs.iter().enumerate() {
+            if let TransactionPayload::TenureChange(..) = &tx.payload {
+                ret.push(i);
+            } else {
+                break;
+            }
+        }
+        ret
+    }
+
+    /// Does this block contain one or more well-formed and valid tenure change transactions?
+    /// Return Some(true) if it does contain at least one, and they're all valid
+    /// Return Some(false) if it does contain at least one, but at least one is invalid
+    /// Return None if it contains none.
+    pub fn tenure_changed(&self) -> Option<bool> {
+        let wellformed = self.is_wellformed_first_tenure_block();
+        if wellformed.is_none() {
+            // block isn't a first-tenure block, so no valid tenure changes
+            return None;
+        } else if let Some(false) = wellformed {
+            // this block is malformed
+            info!("Block is malformed";
+                  "block_id" => %self.block_id());
+            return Some(false);
+        }
```

Comment:
If a function returns an `Option`, you can use the `?` operator with `Option` types:

```suggestion
        if !self.is_wellformed_first_tenure_block()? {
            // this block is malformed
            info!("Block is malformed";
                  "block_id" => %self.block_id());
            return Some(false);
        }
```